iOS/Flows/Conversation/Cells/MessageSequenceCollectionViewDataSource.swift

```diff
@@ -127,24 +152,69 @@ extension MessageSequenceCollectionViewDataSource {
             cell.configure(with: message, showAuthor: false)
         }
     }
+
+    static func createLoadMoreCellRegistration() -> LoadMoreCellRegistration {
+        return LoadMoreCellRegistration { cell, indexPath, item in }
+    }
 }
 
 // MARK: - TimelineCollectionViewLayoutDataSource
 
 extension MessageSequenceCollectionViewDataSource: TimeMachineCollectionViewLayoutDataSource {
 
-    func getTimeMachineItem(forItemAt indexPath: IndexPath) -> TimeMachineLayoutItem? {
-        guard let item = self.itemIdentifier(for: indexPath) else { return nil }
-        let messageController = ChatClient.shared.messageController(cid: item.channelID,
-                                                                    messageId: item.messageID)
+    func getTimeMachineItem(forItemAt indexPath: IndexPath) -> TimeMachineLayoutItemType {
+        guard let item = self.itemIdentifier(for: indexPath) else {
+            return TimeMachineLayoutItem(sortValue: .greatestFiniteMagnitude, shouldShow: false)
+        }
 
-        return messageController.message
+        return self.getTimeMachineItem(forItem: item)
+    }
+
+    private func getTimeMachineItem(forItem item: ItemType) -> TimeMachineLayoutItemType {
+        switch item {
+        case .message(let channelID, let messageID):
+            let messageController = ChatClient.shared.messageController(cid: channelID, messageId: messageID)
+            // If the item doesn't correspond to an actual message, then assume it's a placeholder and
+            // put it at the front.
+            return messageController.message ?? TimeMachineLayoutItem(sortValue: .greatestFiniteMagnitude,
+                                                                      shouldShow: false)
+        case .loadMore:
+            // Get all of the message items.
+            let timeMachineItems: [TimeMachineLayoutItemType]
+            = self.snapshot().itemIdentifiers.compactMap { itemIdentifier in
+                switch itemIdentifier {
+                case .message:
+                    return self.getTimeMachineItem(forItem: itemIdentifier)
+                case .loadMore:
+                    return nil
+                }
+            }
+
+            // Find the oldest message item.
+            guard let oldestItem = timeMachineItems.min(by: { (timeMachineItem1, timeMachineItem2) in
+                return timeMachineItem1.sortValue < timeMachineItem2.sortValue
+            }) else {
+                return TimeMachineLayoutItem(sortValue: -.greatestFiniteMagnitude,
+                                             shouldShow: true)
+            }
+
+            // Put the load more item right before the oldest message item.
+            return TimeMachineLayoutItem(sortValue: oldestItem.sortValue.nextDown, shouldShow: true)
+        }
     }
 }
 
-extension Message: TimeMachineLayoutItem {
+private struct TimeMachineLayoutItem: TimeMachineLayoutItemType {
+    var sortValue: Double
+    var shouldShow: Bool
```

Comment:
Now the datasource can be more explicit about whether an item should be visible. This is mainly for the placeholder message.